src/ngx_http_lua_ssl_session_fetchby.c

```diff
@@ -370,17 +370,40 @@ int
 ngx_http_lua_ssl_client_hello_handler(ngx_ssl_conn_t *ssl_conn,
     int *al, void *arg)
 {
-    int                len;
-    ngx_int_t          rc;
-    const u_char      *id;
+    size_t                       len;
+    ngx_int_t                    rc;
+    const u_char                *id;
+    ngx_connection_t            *c;
+    ngx_http_lua_ssl_ctx_t      *cctx;
 
-    len = SSL_client_hello_get0_session_id(ssl_conn, &id);
+    c = ngx_ssl_get_connection(ssl_conn);
 
-    if (len <= 0) {
-        return SSL_CLIENT_HELLO_SUCCESS;
-    }
+    cctx = ngx_http_lua_ssl_get_ctx(c->ssl->connection);
 
-    rc = ngx_http_lua_ssl_sess_fetch_helper(ssl_conn, id, len);
+    if (cctx && cctx->entered_sess_fetch_handler) {
+        /* not the first time */
+        rc = ngx_http_lua_ssl_sess_fetch_helper(ssl_conn, NULL, 0);
+
+    } else {
+        if (SSL_client_hello_get0_legacy_version(ssl_conn) > SSL3_VERSION) {
+            if (!(SSL_get_options(ssl_conn) & SSL_OP_NO_TICKET)
+                && SSL_client_hello_get0_ext(ssl_conn,
```

Comment:
Handshake Protocol: Client Hello
    Handshake Type: Client Hello (1)
    Length: 227
    Version: TLS 1.2 (0x0303)
    Random: a19ee48afa8bbb92977d2d0758209301bfef9036789d5ccb...
    Session ID Length: 32
    Session ID: f30e59969ebabeb836ad262f9e4fef379da0c1e425bc9b28...
    Cipher Suites Length: 8
    Cipher Suites (4 suites)
    Compression Methods Length: 1
    Compression Methods (1 method)
    Extensions Length: 146
    Extension: server_name (len=17)
    Extension: ec_point_formats (len=4)
    Extension: supported_groups (len=12)
    Extension: SessionTicket TLS (len=0)
    Extension: encrypt_then_mac (len=0)
    Extension: extended_master_secret (len=0)
    Extension: signature_algorithms (len=30)
    Extension: supported_versions (len=3)
    Extension: psk_key_exchange_modes (len=2)
    Extension: key_share (len=38)

TLS 1.3 send random session id (useless), and zero-length SessionTicket, it uses key_share for negotiation and session reuse

It is not a good idea that only check TLSEXT_TYPE_session_ticket, perhaps it is better that check TLSEXT_TYPE_supported_versions to identify whether TLS 1.3 or not

Reply:
@crasyangel
Thanks for the idea for checking `TLSEXT_TYPE_supported_versions`. Although I think it is enough to detect TLS 1.3 resumption via checking the PSK.

This pull request doesn't plan to make `ssl_session_fetch_by_lua*` work (or not work) in TLS 1.3.
To avoid shipping a hack to the master branch, I have removed the PSK check in the previous commit. This pull request only try to solve the problem that `ssl_session_fetch_by_lua*` is incorrectly invoked when the session can be resumed via session ticket.

We will solve the TLS 1.3 problem in another pull request.

Reply:
The PSK check is not enough for TLS 1.3, it would send random session id even when full handshake, and no PSK ext.

Reply:
> The PSK check is not enough for TLS 1.3, it would send random session id even when full handshake, and no PSK ext.

+1

> It is not a good idea that only check TLSEXT_TYPE_session_ticket, perhaps it is better that check TLSEXT_TYPE_supported_versions to identify whether TLS 1.3 or not

I think this would not work either. It seems that TLSEXT_TYPE_supported_versions is parsed after client_hello_cb() called.

Reply:
Yes, TLSEXT_TYPE_supported_versions is parsed after client_hello_cb() called. But we could use  SSL_client_hello_get0_ext to get TLSEXT_TYPE_supported_versions, Client Hello has been processed in tls_process_client_hello(), although openssl assigns ssl member in tls_early_post_process_client_hello(), such as s->version, which executes after client_hello_cb()

And, TLSEXT_TYPE_supported_versions would be a raw string which has packet prefix before client_hello_cb(), but perhaps it is enough to verify if supported_versions exists, as:
`
suppversions = &s->clienthello->pre_proc_exts[TLSEXT_IDX_supported_versions];
if (suppversions->present)
    return;
`

Maybe TLSEXT_TYPE_key_share is a better choice

Reply:
You mean `if (client_support_tls1.3) skip;` or `if (client_support_tls1.3 and server_support_tls1.3) skip;`?
The presence of TLSEXT_TYPE_supported_versions or TLSEXT_TYPE_key_share just indicates the client support TLS1.3, the server still can choose not to use TLS1.3. And a TLS1.3 client may carry a real session id, from a previous TLS1.2 server for example, and the server should get a chance to reuse it.

Reply:
@crasyangel @yongjianchn
For `TLSEXT_TYPE_supported_versions`, We can check if the server supports TLS 1.3 or not.

For the previous session id, maybe the client will also populate it in the PSK extension? I can not find OpenSSL does  resumption work with the legacy client id in TLS 1.3 yet. (Please correct me if I am wrong).

This pull request is not relative with TLS 1.3. Please move to issue https://github.com/openresty/lua-nginx-module/issues/1539 for further discussion.

Reply:
If a TLS1.3 client connected to a TLS1.2 server, for the next connection, it will carry the previous session id in TLS1.2 style. If not, how can the TLS1.2 server do session resumption? But yes, if we also check the server supports TLS1.3 or not, the problem would not exist.

Reply:
@yongjianchn

> And a TLS1.3 client may carry a real session id, from a previous TLS1.2 server for example, and the server should get a chance to reuse it.

If we know the TLS 1.3 won't be selected, we already get a change to reuse it. The `ssl_session_fetch_by_lua*` still works under TLS 1.2.

Reply:
Indeed, you are right, session id would be useful when server does not support TLS 1.3